Изменения магазина и меню

Tanya-g99 opened the pull request:
closes #47

viacheslavek committed:
Заполнил магазин фотографиями и улучшил левое меню

viacheslavek committed:
Добавил названия предметов и поле "Облик"

geogreck committed:
Собрал все предметы в отдельный `json` файл

Tanya-g99 committed:
Внесены изменения в меню магазина

Tanya-g99 committed:
Merge branch 'main' into filling-store-with-items

Tanya-g99 committed:
Внесены изменения в общее меню

viacheslavek committed:
Улучшил заголовок страницы
-добавил название
-добавил иконку
-при нажатии на заголовок раздела нас отсылает к началу

viacheslavek committed:
Создал структуру js файлов магазина

-папка "компоненты магазина" и папка "константы"
-добавил в файл каталог для теста некоторое количество товаров (позже они будут удалены и все будет забираться с файла .json)
-вывожу их в консоль

viacheslavek reviewed (changes requested):
Я ещё пока работаю в этой ветке (делаю подгрузку магазина с json файла), так что пока рано ее закрывать)

viacheslavek committed:
Начал создание магазина как подгрузка с файла js
-Очень сыро подгружается раздел "Одежда"
(то ли еще будет :3 )

viacheslavek committed:
Добавил расшифровку редкости карточки
минус: появился баг с прогрузкой картинок

geogreck commented:
Каков замысел пушить такие файлы?
![Screenshot_2022-08-01-18-19-10-82_21da60175e70af211acc4f26191b7a77](https://user-images.githubusercontent.com/74151984/182182929-9fb9a0a7-8b55-4da1-9760-f4156aff03b6.jpg)

geogreck commented:
Аналогично
![Uploading Screenshot_2022-08-01-18-20-12-91_21da60175e70af211acc4f26191b7a77.jpg…]()

geogreck commented:
Зачем делать копию уже существующего файла
![Screenshot_2022-08-01-18-20-03-57_21da60175e70af211acc4f26191b7a77](https://user-images.githubusercontent.com/74151984/182183209-a64ffc86-c94f-4978-a42d-a90bdede7c62.jpg)


Позже добавлю поддержку json
(Пока разбирался с тем, как работает вставка в html файл поля из объектов js)

geogreck commented:
Аналогично
![Screenshot_2022-08-01-18-20-12-91_21da60175e70af211acc4f26191b7a77](https://user-images.githubusercontent.com/74151984/182183361-1f341667-bee1-4fb0-ab9b-29cf1249e5ed.jpg)

viacheslavek commented:
В последующих коммитах удалю пустые файлы)
Сначала я задумал одну архитектуру, но потом решил обойтись единым js файлом.

geogreck commented:
А смысл пушить настолько кривую, не рабочую и ужасно оформленную версию

viacheslavek committed:
Произвел рефактор файлов:
-добавил картинки предметов в папку assets
-удалил пустые файлы

viacheslavek committed:
Добавил fetch запрос для получения информации о предметах с сервера
-соответсвенно, удалил временный js файл каталог
-сделал правки в json файл как на сервере

viacheslavek committed:
Дописал формирование заголовков для разных типов предметов

viacheslavek committed:
Убрал однотипные блоки кода снизу и формирование блоков заголовка по id предметов (надо придумать, как закрывать блоки без использования id)

viacheslavek committed:
Изменил теги заголовков под документацию и соеденил блок создания заголовков

viacheslavek commented:
Уже менее кривая, более рабочая и вроде бы средне оформленная версия 😅

viacheslavek committed:
Внёс правки в json файл
- исправил баги с названием картинок (кое-какие стоит перименовать)
- добавил редкост в карточки

geogreck commented:
Стыдно! Красиво называйте свои коммиты.
![Screenshot_2022-08-03-08-18-36-91_21da60175e70af211acc4f26191b7a77](https://user-images.githubusercontent.com/74151984/182529771-b9f1f32b-fcf6-4f9f-9ca6-efcd887d1777.jpg)

viacheslavek committed:
Теперь вместо js использую ts (дальше только на нём).
-подправил скрипт магазина под ts
-изменил в json файле тип одежды шлем на "helmet"
-добавил функцию для окрашивания цвета карточки (Таня, спасибо, взял у тебя :)) )
-убрал id для определения конца блока магазина

viacheslavek committed:
Добавлены модальные окна.
Открываются при нажатии на кнопку "купить" в магазине
-теперь описание будет содержаться только в модальном окне

viacheslavek committed:
Добавлена проверка на "купленность".
-добавлена строка "баланс" в верхнем меню (позже будет получать информацию с сервера)
-если предмет куплен, его нельзя купить вновь
-удален старый магазин, ведь на сервер почти все загрузили

viacheslavek commented:
Позже добавлю построение модальных окон, используя dom дерево

viacheslavek committed:
Добавлены DOM модальные окна.
ура)

viacheslavek commented:
Сделать модальное окно по центру

geogreck commented on frontend/js/componentsShop/dom_modal.ts:
```diff
@@ -0,0 +1,103 @@
+type Item = {
+    id: number;
```
Каков смысл существования этого файла, если он содержит уже только написанный Таней код ? Я бы сказал, 
"Обнаружены ОЧЕНЬ похожие решения"

geogreck commented on frontend/js/componentsShop/products.ts:
```diff
@@ -0,0 +1,102 @@
+    if (rarity === "common") return "#C8C8C8";
+    else if (rarity === 'rare') return "#FFB74D";
+    else if (rarity === 'epic') return "#26A69A";
+    else if (rarity === 'legendary') return "#F06292";
```
Эта функция копирует функцию, которая была скопирована у Татьяны в другом файле

geogreck commented on frontend/js/componentsShop/dom_modal.ts:
```diff
@@ -0,0 +1,103 @@
+
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore
+const modalInventory = new bootstrap.Modal(<HTMLFormElement>document.getElementById('shopModal'));
```
Код настолько внимательно копировался у Татьяны, что не изменено даже название модалки :)

geogreck replied:
Кстати в продакшене это будет приводить к ошибке выполнения :)

geogreck commented:
Не помешало бы убрать код, который уже добавлен Татьяной при создании инвентаря, чтобы не было дублирования + убрать все `ts-ignore`

viacheslavek commented:
По твоим комментариям сверху может сложиться впечатление, что я втихую присвоил код Татьяны. Но ведь мы с тобой вместе обсуждали то, что проще и быстрее будет  использовать уже имеющуюся технологию, и для этого Таня помогала мне, добавив комментарии и отвечав мне на мои вопросы в сообщениях. Функцию получения цвета я пока оставил здесь, так как она находится в другой ветке и пока мне недоступна, как и тип item, позже, после слияния, мы оставим только одну. Остальные функции я подстраивал под то, что мне нужно на этой странице. Замечания принял к сведению и исправлю их. 
В некоторых ваших комментариях я хотел бы лучше видеть анекдоты, а не попытку задеть меня(

geogreck commented:
То что код бывает реализован в другой ветке, а в твоей нет, это обычное дело. Есть два простых пути решения:
1) (так себе) слить ветку с кодом в свою ветку
2) (хороший) пнуть авторов другой ветки, чтобы ее слили в Мейн и уже слить мейн в свою ветку
А пушить такой код ещё раз в свою ветку никак нельзя, так как это банально приведет к неработоспособности проекта в мейне, что можно не отловить заранее, особенно учитывая то, что у тебя весь код в `@ts-ignore`.

viacheslavek committed:
Добавил загрузку баланса.
(с тестового сервера)

viacheslavek committed:
Добавлена возможность покупать предмет.
-Осталось отправлять баланс на сервер (тестовый), а предмет в инвентарь

viacheslavek committed:
Правки внешнего вида.
-модальные окна всегда по центру
-кнопки купить сверху без закруглений

viacheslavek committed:
Произведен рефакторинг.
-убраны дублирующие функции
-убраны ts-ignore
-добавлена ссылка на инвентарь

viacheslavek commented:
Основная работа сделана, буду исправлять замечания (предписания) Георгия

geogreck reviewed (commented):
К правкам Татьяны и картинкам вопросов не имею

geogreck commented on frontend/html/shop.html:
```diff
@@ -4,18 +4,31 @@
+    <script src="../js/request.ts"></script>
+    <script defer src="../js/config.ts"></script>
+    <script defer src="../js/list/utility.ts"></script>
+    <script defer src="../js/inventory/inventory.ts"></script>
```
Зачем подключается этот скрипт?

viacheslavek replied:
Для функции определения цвета карточки и доступа к типу Item (последние две)

viacheslavek replied:
Или это можно сделать как-то лучше?)

geogreck replied:
Ничего из вышеперечисленного в файле `inventory.ts` нет

viacheslavek replied:
Упс

geogreck commented on frontend/html/shop.html:
```diff
@@ -24,346 +37,127 @@
-                </div>
-            </div>
         </div>
+        <br>
```
Зачем `<br>`?

viacheslavek replied:
![Снимок 10 08 2022 в 12 21](https://user-images.githubusercontent.com/89520622/183865503-ed8e1b61-edc8-4f9c-8dba-6a5ba45ee212.jpg)
с <br> проще попасть на кнопку купить, да и просто глазу приятнее
вот без
![Снимок 10 08 2022 в 12 21](https://user-images.githubusercontent.com/89520622/183865652-66d9bd76-ede5-4922-a5ee-3fba661203a3.jpg)
:

viacheslavek replied:
На белом фоне не видно...... Ну тогда стоит поверить на слово (или проверить :3)

geogreck replied:
Для этого используют свойства `margin` или `padding`

viacheslavek replied:
Изучил - согласен

geogreck commented on frontend/html/shop.html:
```diff
@@ -24,346 +37,127 @@
+            Редкая:      #FFB74D ($orange-300)
+            Эпическая:   #F06292 ($pink-300)
+            Легендарная: #26A69A ($teal-400)
+            Мифическая:  linear-gradient(#40E0D0, #FF8C00, #FF0080)
```
Ненужный комментарий

geogreck commented on frontend/js/componentsShop/dom_modal.ts:
```diff
@@ -0,0 +1,60 @@
+const itemsShop = new Map<number, Item>();
```
Файл неплохо было бы прогнать через форматтер, слишком уж много лишних отступов

viacheslavek replied:
Окс

viacheslavek replied:
Прогнал через встроенный в ide, некоторые добавил сам для читабельности (разделения на блоки)

geogreck commented on frontend/js/componentsShop/dom_modal.ts:
```diff
@@ -0,0 +1,60 @@
+    // получение из хранилища предмета, на карту которого кликнули
+    const regexp = /idItem=\d+/;
+
+    const strId = findID(target, regexp);
```
Снова лишний отступ

geogreck commented on frontend/js/componentsShop/dom_modal.ts:
```diff
@@ -0,0 +1,60 @@
+        footer.innerHTML = `
+            <button type="button" class="buyButton btn btn-primary btn-lg">
+                Купить за ${item.price} todoкоинов?
+            </button>
```
Кнопка в модалке не по центру

geogreck commented on frontend/js/componentsShop/dom_modal.ts:
```diff
@@ -0,0 +1,60 @@
+        const windowForm = <HTMLElement>document.getElementsByClassName('shop__form')[0];
+        const buf = (<HTMLElement>document.getElementsByClassName(`idItem=${id}`)[0]).getBoundingClientRect();
+
+        // делать по центру одинаково
```
По вертикали модальное окно не по центру

geogreck commented on frontend/js/componentsShop/products.ts:
```diff
@@ -0,0 +1,170 @@
+        const namingBlock: string[] = ['Одежда', 'Аксессуары', 'Питомцы', 'Облик'];
+        let previousCategory = 'helmet';
+
+        CATALOG_SHOP.forEach(({id, name, description, imageSrc, category, rarity, state}) => {
```
Почему не используется просто элемент типа `item`

geogreck commented on frontend/js/componentsShop/products.ts:
```diff
@@ -0,0 +1,170 @@
+class Products {
```
Файл бы тоже прогнать через форматтер

geogreck commented on frontend/js/componentsShop/products.ts:
```diff
@@ -0,0 +1,170 @@
+    const balanceShop = <HTMLInputElement>document.getElementById('balance');
+    balanceShop.innerText = 'Ваш баланс: ' + balance.toString() + ' коинов';
+
+})
```
Почему баланс хранится только в коде страницы?

geogreck commented on frontend/js/componentsShop/products.ts:
```diff
@@ -0,0 +1,170 @@
+    if (hasParentClass(targetBuy, 'buyButton')) {
+        const costAndString: string = targetBuy.innerHTML;
+        const cost: number = parseInt(costAndString.replace(/\D+/g,""));
+        const itemBuy: Item = itemsCostShop.get(cost)!;
```
Довольно странно каждый раз производить такие длинные вычисления, вместо того, чтобы просто взять цену из таблицы предметов

geogreck commented on frontend/js/componentsShop/products.ts:
```diff
@@ -0,0 +1,170 @@
+sendRequest('GET', server + '/items').then((data) => {
+    for (let i = 0; i < data.length; i++) {
+        const item: Item = data[i];
+        itemsCostShop.set(item.price, item);
```
Замысел этого кода к сожалению понять не удалось. Зачем делать ещё одно хранилище предметов, если есть itemsShop?

geogreck commented on frontend/js/componentsShop/products.ts:
```diff
@@ -0,0 +1,170 @@
+
+        const balanceShopForBuy: HTMLInputElement = <HTMLInputElement>document.getElementById('balance');
+        const balanceShopForBuyAndString: string = balanceShopForBuy.innerHTML;
+        let balanceBuy: number = parseInt(balanceShopForBuyAndString.replace(/\D+/g,""));
```
Довольно странно каждый раз производить такие длинные вычисления, вместо того, чтобы просто взять баланс из переменной

geogreck commented on frontend/js/componentsShop/products.ts:
```diff
@@ -0,0 +1,170 @@
+        let balanceBuy: number = parseInt(balanceShopForBuyAndString.replace(/\D+/g,""));
+
+        // меняем состояние
+        console.log(balanceBuy);
```
Не убранный console.log

geogreck commented on frontend/js/componentsShop/products.ts:
```diff
@@ -0,0 +1,170 @@
+            sendRequest('PATCH', server + '/users/' + user, JSON.stringify({ balance: balanceBuy }))
+            sendRequest('PATCH', server + `/items/${itemBuy.id}`, JSON.stringify({ state: itemBuy.state }))
+            console.log(balanceBuy);
+            console.log(itemBuy);
```
Снова они же

geogreck commented on frontend/js/componentsShop/products.ts:
```diff
@@ -0,0 +1,170 @@
+    })
+    .catch(error => {
+        console.log(error);
+    })
```
Для этого так же можно  использовать функцию `sendRequest`

geogreck commented on frontend/js/componentsShop/products.ts:
```diff
@@ -0,0 +1,170 @@
+            console.log(balanceBuy);
+            console.log(itemBuy);
+        }
+        else if (balanceBuy >= cost && (itemBuy.state == 'inventoried' || itemBuy.state == 'equipped')) {
```
Очень странно написаны `if`-ы. Во-первых если верить надежности кода, у купленного предмета изначально написано, что его купить нельзя, а значит смысл от этих всех проверок? Во-вторых если вдруг нет уверенности в работоспособности своего кода и такая проверка нужна, можно сначала сделать проверку `balanceBuy >= cost` и уже внутри сделать проверку, не куплен ли предмет

geogreck commented on frontend/js/componentsShop/products.ts:
```diff
@@ -0,0 +1,170 @@
+
+    render() {
+        let htmlCatalog = '';
+        let numberBlock = 0;
```
Зачем нужна переменная `numberBlock`

geogreck commented on frontend/js/componentsShop/products.ts:
```diff
@@ -0,0 +1,170 @@
+                        </div>
+                            ${buildingBuyButton(id, state)}
+
+                        <!- модальное окно --->
```
Не убранный комментарий

geogreck commented on frontend/js/componentsShop/products.ts:
```diff
@@ -0,0 +1,170 @@
+            if ((category === 'armor' || category === 'helmet') && numberBlock === 0 ||
+                category === 'weapon' && numberBlock === 1 ||
+                category === 'pet' && numberBlock === 2 ||
+                category === 'skin' && numberBlock === 3) {
```
ОЧЕНЬ странный `if`. Если уж выбран странный алгоритм запихивания всего каталога в одну переменную, то проверку на новый логический блок можно делать по смене типа предмета

geogreck commented on frontend/js/componentsShop/products.ts:
```diff
@@ -0,0 +1,170 @@
+                    htmlCatalog += `
+                            </div>
+                        </div>
+                        <br>
```
Зачем `<br>`?

geogreck commented on frontend/js/componentsShop/products.ts:
```diff
@@ -0,0 +1,170 @@
+
+        const html = `
+                ${htmlCatalog}
+        `;
```
Что делает этот код?

geogreck replied:
вопрос риторический - ничего

geogreck commented on frontend/js/componentsShop/products.ts:
```diff
@@ -0,0 +1,170 @@
+        `;
+
+        const ROOT_PRODUCTS = document.getElementById('products');
+        ROOT_PRODUCTS.innerHTML = html;
```
Весь код страдает от разного стиля названия переменных, то используется `camelCase`, то `CONSTATN_CASE`, причем не только для констант

geogreck commented on frontend/js/componentsShop/products.ts:
```diff
@@ -0,0 +1,170 @@
+
+const productsPage = new Products();
+
+let CATALOG_SHOP = [];
```
Как тут

geogreck commented on frontend/js/componentsShop/products.ts:
```diff
@@ -0,0 +1,170 @@
+            <button type="button" class="for__click btn btn-success disabled idItem=${id}" data-bs-toggle="modal"
+                                data-bs-target="#selling${id}" id="buttonBuy${id}"
+                                style="border-radius: 0 0 3px 3px;">
+                                Куплено</button>
```
А зачем это создавать как активную кнопку, если предмет уже куплен и ничего с ним сделать нельзя?

viacheslavek replied:
Изменил систему кнопок
см - первый кормит после ревью

geogreck commented on frontend/js/componentsShop/dom_modal.ts:
```diff
@@ -0,0 +1,60 @@
+        `;
+
+    }
+});
```
Нет перехода на новую строку. Чинится запуском форматтера.

geogreck commented:
Забыл указать, что открытие любого модального окна приводит к ошибке:
```js
Uncaught TypeError: Cannot read properties of undefined (reading 'classList')
    at Hi._isAnimated (modal.js:313:26)
    at Hi._initializeBackDrop (modal.js:195:24)
    at new Hi (modal.js:83:27)
    at Function.getOrCreateInstance (base-component.js:55:41)
    at HTMLButtonElement.<anonymous> (modal.js:421:22)
    at HTMLDocument.n (event-handler.js:119:21)
```

geogreck commented:
И почему при покупке предмета об этом никак не пишется в модальном окне, нужно обновлять страницу? 
Это не так уж и сложно

viacheslavek commented:
Да, это нужно добавить

viacheslavek committed:
Добавлены красивые кнопки.
-при покупке предмета кнопка сразу зеленеет в окне и становится неактивной
-сразу зеленеет в самом магазине
-после покупки в магазине зеленая кнопка активная, чтобы увидеть описание (можно будет добавить возможность смотреть на цену, за которую был куплен предмет)

viacheslavek committed:
Внес правки Гоши:
-убрал лишние комментарии
-форматнул
-убрал лишний код (html = htmlCatalog)
-переменные названы в единном стиле

Правки-правки-правки
-баланс так же сразу изменяется при покупке (без перезагрузки страницы)
-исправлен баг с кнопкой купить (после нажатия на купленный предмет следующее модальное окно содержало две кнопки)
-решена проблема с ошибкой при нажатии на модальное окно
-элемнт при покупке получаю по id
-убран br - заменен на padding

viacheslavek committed:
Правки-правки-правки-правки
-кнопка покупки в модалке по центру
-модальное окно выровнено в высоту
-баланс берется с сервера
-мелкие изменения в конструкциях if-ов
-Добавил onLoad функцию
-fetch заменен на sendRequest
-каталог заполняю через поля предмета Item

viacheslavek commented:
Вроде бы все правки внесены, жду вашего решения, @geogreck :)

geogreck commented on frontend/js/componentsShop/dom_modal.ts:
```diff
@@ -0,0 +1,46 @@
+        const buf = (<HTMLElement>document.getElementsByClassName(`idItem=${id}`)[0]).getBoundingClientRect();
+
+        // делать по центру одинаково
+        windowForm.style.width = `calc(${buf.right - buf.x}*1.2px + 3vw + 3.8rem)`;
```
Бесполезный код, не делающий ничего

geogreck commented:
Мюллер решил устроить проверку сотрудников. Вызывает в кабинет Шелленберга:
— Шелленберг, назовите какое-нибудь число. Любое.
— Ну, предположим, 32.
— А почему не 23?
— потому что 32!
Мюллер пишет в личном деле: «Характер нордический, стойкий».
Вызывает Кальтенбруннера.
— Кальтенбруннер, назовите какое-нибудь число.
— 67.
— А почему не 76?
— Ну, можно и 76.
Мюллер пишет в личном деле: «Характер близкий к нордическому». Вызывает следующего.
— geogreck closed this 19 hours ago
— VyacheslavIsWorkingNow reopened this 17 hours ago!
— А почему не... А-а, это вы, @VyacheslavIsWorkingNow !

geogreck committed:
Внесены изменения:
- Поправлено отображение баланса
- Добавлено закрытие модального окна в магазине
- Улучшена мобильная версия магазина
- Добавлена навигация по странице с помощью меню магазина
- Немного переименовал файлы и папки

Co-authored-by: Tanya Gnatenko <108430123+Tanya-g99@users.noreply.github.com>
